### DISCUSSION TITLE
Add option to sort by random in collection list response

### DISCUSSION THREAD

**Discussion by maxicapodacqua** [pocketbase/pocketbase#1393] [Q&A]
Currently there is no option to retrieve a list of collection items sorted by random. Since sqlite allows sorting by `random()` it should be possible to implement this in the backend.

I don't have any recommendation on the name of the field for the Web API, but I'm currently working on a project that is going to need this sort feature. It can be implemented from the client by using random page numbers, but I'd like to get sets of rows directly from the backend as is better for performance.

Any recommendations on the implementation or field names to use?

**Comment by ganigeorgiev** [comment]
Hm, I don't think we can add this since `getList()` returns paginated results and I'm not sure how it will work if we sort by random every time it is called.

As you noted, you can emulate the "randomness" client-side. I don't think there will be too much difference in the performance.

If you need to execute custom query, you can [extend PocketBase](https://pocketbase.io/docs/use-as-framework/), [register a new API endpoint](https://pocketbase.io/docs/router/#register-new-route) and use the [internal db abstraction](https://pocketbase.io/docs/database/).

_Eventually in the future we may add a "view" collection type and/or "raw sql" field type per https://github.com/pocketbase/pocketbase/issues/311#issuecomment-1312790397 that will allow specifying a custom query from the Admin UI._

**Reply by maxicapodacqua** [reply]
Agree that it can't be implemented for pagination, but I think the performance is different if I emulate randomness client-side if the use case is to get a list of random elements.
For example, for my project I need to get 100 random elements and show those to the user, if I need to do that using client side and get a random page number I will have to run 100 fetch of single elements using random page numbers.

I'm investigating how to do that using an extra column that can generate a different sort result

**Reply by ganigeorgiev** [reply]
It may not be suitable for your use case, but if you don't want to execute a custom query, one hacky workaround could be to fetch a single random page with slightly more items (eg. 200) and shuffle the result client side returning only the first 100.

**Comment by Fish1** [comment]
I was also just looking for this feature... I have a database where I might have thousands of posts, and I'd like a feature where I can query a random one to show to users.

**Comment by ganigeorgiev** [comment]
@maxicapodacqua, @Fish1

I've added an exception for a special `@random` sort key as a workaround in the `filter-enhancements` branch (it will be available sometime next week with v0.11.0). Or in other words, you'll be able to call:

```js
// /api/collections/example/records?perPage=50&sort=@random
const result = await pb.collection('example').getList(1, 50, { sort: "@random" });
```
_(obviously since the order will be random, the offset and cursor based pagination will not work; it is also still possible to combine the `@random` sort key with other sort fields, for example `?sort=title,@random` and in this case the randomness will have effect only on records with duplicated `title`)_

**Comment by leikoilja** [comment]
@ganigeorgiev, has `@random` ever been released?

**Reply by ganigeorgiev** [reply]
@leikoilja Yes, it should be supported (there is also a unit test for it in https://github.com/pocketbase/pocketbase/blob/b7407edcae15c469f00ef3caf6e6da6cf0d30876/tools/search/sort_test.go#L30-L31).

**Reply by leikoilja** [reply]
Great, sorry, missed the docs in https://pocketbase.io/docs/api-records/. Works smooth :)
